src/frontend/js/widgets/SyllabusCourseRunsList/components/CourseProductItem/index.spec.tsx

```diff
@@ -586,6 +586,65 @@ describe('CourseProductItem', () => {
     expect(screen.queryByTestId('PurchaseButton__cta')).toBeNull();
   });
 
+  it('renders a button and a banner if the contract needs to be signed', async () => {
+    const relation = CourseProductRelationFactory().one();
+    const { product } = relation;
+    const order = CredentialOrderFactory({
+      product: product.id,
+      target_courses: product.target_courses,
+      course: CourseLightFactory({ code: '00000' }).one(),
+      state: OrderState.SUBMITTED,
+    }).one();
+    fetchMock.get(`https://joanie.test/api/v1.0/courses/00000/products/${product.id}/`, relation);
+    const orderQueryParameters = {
+      product: order.product,
+      course: order.course.code,
+      state: ACTIVE_ORDER_STATES,
+    };
+
+    fetchMock.get(
+      `https://joanie.test/api/v1.0/orders/?${queryString.stringify(orderQueryParameters)}`,
+      [order],
+    );
+
+    render(
+      <Wrapper withSession>
+        <CourseProductItem productId={product.id} courseCode="00000" />
+      </Wrapper>,
+    );
+
+    // Wait for product information to be fetched
+    const loadingMessage = screen.getByRole('status', { name: 'Loading product information...' });
+    await waitForElementToBeRemoved(loadingMessage);
+    await screen.findByRole('heading', { level: 3, name: product.title });
+
+    // - A banner should be displayed.
+    screen.getByText('You need to sign your contract before enrolling to course runs');
+
+    screen.getByRole('link', { name: 'Go to signature' });
+
+    // - In place of product price, a label "Pending" should be displayed
+    const $enrolledInfo = await screen.findByText('Pending');
+    expect($enrolledInfo.tagName).toBe('STRONG');
+    expect($enrolledInfo.classList.contains('h6')).toBe(true);
+
+    // - As order is pending, the user should not be able to enroll to course runs.
+    await waitFor(() => {
+      order.target_courses.forEach((course) => {
+        const $item = screen.getByTestId(`course-item-${course.code}`);
+        // the course title shouldn't be a heading to prevent misdirection for screen reader users,
+        // but we want to it to visually look like a h5
+        const $courseTitle = getByText($item, course.title);
+        expect($courseTitle.tagName).toBe('STRONG');
+        expect($courseTitle.classList.contains('h5')).toBe(true);
+        screen.getByTestId(`CourseRunList-${course.course_runs.map(({ id }) => id).join('-')}`);
+      });
+    });
+
+    // - Does not Render PurchaseButton cta
+    expect(screen.queryByTestId('PurchaseButton__cta')).toBeNull();
+  });
```

Comment:
That's quite bad, every n second `waitFor` will re-execute it's callback.
Here we'll rerun the `forEach` x times until its get resolved.

I didn't like that we loose the  `expect().toBeInTheDocument()` but something like this would do the job:
```suggestion
    await Promise.all(
      order.target_courses.map((course) =>
        screen.findByTestId(`CourseRunList-${course.course_runs.map(({ id }) => id).join('-')}`),
      ),
    );
```